// thirdweb/dist/cjs/react/web/ui/TransactionButton/DepositScreen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DepositScreen = DepositScreen;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const address_js_1 = require("../../../../utils/address.js");
const formatNumber_js_1 = require("../../../../utils/formatNumber.js");
const units_js_1 = require("../../../../utils/units.js");
const is_smart_wallet_js_1 = require("../../../../wallets/smart/is-smart-wallet.js");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const useActiveAccount_js_1 = require("../../../core/hooks/wallets/useActiveAccount.js");
const useActiveWallet_js_1 = require("../../../core/hooks/wallets/useActiveWallet.js");
const ErrorState_js_1 = require("../../wallets/shared/ErrorState.js");
const LoadingScreen_js_1 = require("../../wallets/shared/LoadingScreen.js");
const CoinsIcon_js_1 = require("../ConnectWallet/icons/CoinsIcon.js");
const useBuyTxStates_js_1 = require("../ConnectWallet/screens/Buy/main/useBuyTxStates.js");
const WalletRow_js_1 = require("../ConnectWallet/screens/Buy/swap/WalletRow.js");
const formatTokenBalance_js_1 = require("../ConnectWallet/screens/formatTokenBalance.js");
const nativeToken_js_1 = require("../ConnectWallet/screens/nativeToken.js");
const CopyIcon_js_1 = require("../components/CopyIcon.js");
const QRCode_js_1 = require("../components/QRCode.js");
const Skeleton_js_1 = require("../components/Skeleton.js");
const Spacer_js_1 = require("../components/Spacer.js");
const WalletImage_js_1 = require("../components/WalletImage.js");
const basic_js_1 = require("../components/basic.js");
const buttons_js_1 = require("../components/buttons.js");
const text_js_1 = require("../components/text.js");
const TokenSymbol_js_1 = require("../components/token/TokenSymbol.js");
const elements_js_1 = require("../design-system/elements.js");
const useCopyClipboard_js_1 = require("../hooks/useCopyClipboard.js");
const pulseAnimation = (0, react_1.keyframes) `
0% {
  opacity: 1;
  transform: scale(0.5);
}
100% {
  opacity: 0;
  transform: scale(1.5);
}
`;
const WaitingBadge = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        display: "flex",
        alignItems: "center",
        gap: index_js_1.spacing.sm,
        backgroundColor: theme.colors.tertiaryBg,
        border: `1px solid ${theme.colors.borderColor}`,
        padding: `${index_js_1.spacing.md} ${index_js_1.spacing.sm}`,
        borderRadius: index_js_1.radius.lg,
        color: theme.colors.secondaryText,
        fontSize: index_js_1.fontSize.sm,
        fontWeight: 500,
        position: "relative",
        "&::before": {
            content: '""',
            width: "8px",
            height: "8px",
            borderRadius: "50%",
            backgroundColor: theme.colors.accentText,
            animation: `${pulseAnimation} 1s infinite`,
        },
    };
});
/**
 *
 * @internal
 */
function DepositScreen(props) {
    const activeWallet = (0, useActiveWallet_js_1.useActiveWallet)();
    const activeAccount = (0, useActiveAccount_js_1.useActiveAccount)();
    const address = activeAccount?.address;
    const { hasCopied, onCopy } = (0, useCopyClipboard_js_1.useClipboard)(address || "");
    const { connectLocale, client } = props;
    const locale = connectLocale.receiveFundsScreen;
    const isTestnet = props.tx.chain.testnet === true;
    const { data: transactionCostAndData, error: transactionCostAndDataError, isFetching: transactionCostAndDataFetching, refetch: transactionCostAndDataRefetch, } = (0, useBuyTxStates_js_1.useTransactionCostAndData)({
        transaction: props.tx,
        account: activeAccount,
        supportedDestinations: [],
        refetchIntervalMs: 10_000,
    });
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    const sponsoredTransactionsEnabled = (0, is_smart_wallet_js_1.hasSponsoredTransactionsEnabled)(activeWallet);
    if (transactionCostAndDataError) {
        return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { style: {
                minHeight: "350px",
            }, fullHeight: true, flex: "row", center: "both", children: (0, jsx_runtime_1.jsx)(ErrorState_js_1.ErrorState, { title: transactionCostAndDataError?.message || "Something went wrong", onTryAgain: transactionCostAndDataRefetch }) }));
    }
    if (!transactionCostAndData) {
        return (0, jsx_runtime_1.jsx)(LoadingScreen_js_1.LoadingScreen, {});
    }
    const totalCost = (0, nativeToken_js_1.isNativeToken)(transactionCostAndData.token) && !sponsoredTransactionsEnabled
        ? transactionCostAndData.transactionValueWei +
            transactionCostAndData.gasCostWei
        : transactionCostAndData.transactionValueWei;
    const insufficientFunds = transactionCostAndData.walletBalance.value < totalCost;
    const requiredFunds = transactionCostAndData.walletBalance.value
        ? totalCost - transactionCostAndData.walletBalance.value
        : totalCost;
    const openFaucetLink = () => {
        window.open(`https://thirdweb.com/${props.tx.chain.id}?utm_source=ub_deposit`);
    };
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", children: [(0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: "Insufficient funds", onBack: props.onBack }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "sm", children: [insufficientFunds && ((0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsxs)(text_js_1.Text, { size: "xs", center: true, color: "danger", multiline: true, children: ["You need", " ", (0, formatNumber_js_1.formatNumber)(Number.parseFloat((0, units_js_1.toTokens)(requiredFunds, transactionCostAndData.decimals)), 5), " ", transactionCostAndData.token.symbol, " to continue"] }) })), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", style: {
                            justifyContent: "space-between",
                            padding: index_js_1.spacing.sm,
                            marginBottom: index_js_1.spacing.sm,
                            borderRadius: index_js_1.spacing.md,
                            backgroundColor: theme.colors.tertiaryBg,
                            border: `1px solid ${theme.colors.borderColor}`,
                        }, children: [activeAccount && ((0, jsx_runtime_1.jsx)(WalletRow_js_1.WalletRow, { address: activeAccount?.address, iconSize: "md", client: client })), transactionCostAndData.walletBalance.value !== undefined &&
                                !transactionCostAndDataFetching ? ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", gap: "3xs", center: "y", children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", color: "secondaryText", weight: 500, children: (0, formatTokenBalance_js_1.formatTokenBalance)(transactionCostAndData.walletBalance, false) }), (0, jsx_runtime_1.jsx)(TokenSymbol_js_1.TokenSymbol, { token: transactionCostAndData.token, chain: props.tx.chain, size: "xs", color: "secondaryText" })] })) : ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", gap: "3xs", center: "y", children: (0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { width: "70px", height: index_js_1.fontSize.xs }) }))] })] }), (0, jsx_runtime_1.jsx)(WalletAddressContainer, { onClick: onCopy, children: (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "md", center: "both", expand: true, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", children: (0, jsx_runtime_1.jsx)(QRCode_js_1.QRCode, { qrCodeUri: address, size: 250, QRIcon: activeWallet && ((0, jsx_runtime_1.jsx)(WalletImage_js_1.WalletImage, { id: activeWallet.id, size: index_js_1.iconSize.xl, client: client })) }) }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", center: "x", gap: "xs", children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "md", children: address ? (0, address_js_1.shortenAddress)(address) : "" }), (0, jsx_runtime_1.jsx)(CopyIcon_js_1.CopyIcon, { text: address || "", tip: "Copy address", hasCopied: hasCopied })] })] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { multiline: true, center: true, balance: true, size: "sm", className: "receive_fund_screen_instruction", children: locale.instruction }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), insufficientFunds ? ((0, jsx_runtime_1.jsxs)(WaitingBadge, { children: ["Waiting for funds on ", transactionCostAndData.chainMetadata.name, "..."] })) : ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "accent", onClick: props.onDone, fullWidth: true, children: "Continue" })), insufficientFunds && isTestnet && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "link", onClick: openFaucetLink, fullWidth: true, children: (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", center: "x", gap: "xs", color: "accentText", children: [(0, jsx_runtime_1.jsx)(CoinsIcon_js_1.CoinsIcon, { size: index_js_1.iconSize.sm }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", color: "accentText", weight: 500, center: true, children: "Get testnet funds" })] }) })] }))] }));
}
const WalletAddressContainer = /* @__PURE__ */ (0, elements_js_1.StyledButton)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        all: "unset",
        width: "100%",
        boxSizing: "border-box",
        cursor: "pointer",
        padding: index_js_1.spacing.md,
        display: "flex",
        justifyContent: "space-between",
        border: `1px solid ${theme.colors.borderColor}`,
        borderRadius: index_js_1.radius.lg,
        transition: "border-color 200ms ease",
        "&:hover": {
            borderColor: theme.colors.accentText,
        },
    };
});
