// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/PayProviderSelection.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ChevronDownIcon } from "@radix-ui/react-icons";
import { iconSize, spacing } from "../../../../../core/design-system/index.js";
import { Container } from "../../../components/basic.js";
import { Button } from "../../../components/buttons.js";
import { Text } from "../../../components/text.js";
import { getProviderLabel } from "./utils.js";
/**
 * Shows the selected payment provider based on the preferred provider OR the quoted provider.
 * @internal
 */
export const PayProviderSelection = (props) => {
    const ProviderItem = (_jsxs(Container, { flex: "row", center: "y", gap: "xxs", color: "secondaryText", style: { padding: spacing.md }, children: [_jsx(Text, { size: "xs", children: getProviderLabel(props.preferredProvider ?? props.quotedProvider ?? "") }), props.supportedProviders.length > 1 && (_jsx(ChevronDownIcon, { width: iconSize.sm, height: iconSize.sm }))] }));
    return (_jsxs(Container, { bg: "tertiaryBg", flex: "row", borderColor: "borderColor", style: {
            paddingLeft: spacing.md,
            justifyContent: "space-between",
            alignItems: "center",
            borderWidth: "1px",
            borderStyle: "solid",
            borderBottom: "none",
        }, children: [_jsx(Text, { size: "xs", color: "secondaryText", children: "Provider" }), props.supportedProviders.length > 1 ? (_jsx(Button, { variant: "ghost", onClick: props.onShowProviders, style: { padding: 0 }, children: ProviderItem })) : (ProviderItem)] }));
};
