// thirdweb/dist/esm/react/web/ui/TransactionButton/DepositScreen.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { keyframes } from "@emotion/react";
import { shortenAddress } from "../../../../utils/address.js";
import { formatNumber } from "../../../../utils/formatNumber.js";
import { toTokens } from "../../../../utils/units.js";
import { hasSponsoredTransactionsEnabled } from "../../../../wallets/smart/is-smart-wallet.js";
import { useCustomTheme } from "../../../core/design-system/CustomThemeProvider.js";
import { fontSize, iconSize, radius, spacing, } from "../../../core/design-system/index.js";
import { useActiveAccount } from "../../../core/hooks/wallets/useActiveAccount.js";
import { useActiveWallet } from "../../../core/hooks/wallets/useActiveWallet.js";
import { ErrorState } from "../../wallets/shared/ErrorState.js";
import { LoadingScreen } from "../../wallets/shared/LoadingScreen.js";
import { CoinsIcon } from "../ConnectWallet/icons/CoinsIcon.js";
import { useTransactionCostAndData } from "../ConnectWallet/screens/Buy/main/useBuyTxStates.js";
import { WalletRow } from "../ConnectWallet/screens/Buy/swap/WalletRow.js";
import { formatTokenBalance } from "../ConnectWallet/screens/formatTokenBalance.js";
import { isNativeToken } from "../ConnectWallet/screens/nativeToken.js";
import { CopyIcon } from "../components/CopyIcon.js";
import { QRCode } from "../components/QRCode.js";
import { Skeleton } from "../components/Skeleton.js";
import { Spacer } from "../components/Spacer.js";
import { WalletImage } from "../components/WalletImage.js";
import { Container, ModalHeader } from "../components/basic.js";
import { Button } from "../components/buttons.js";
import { Text } from "../components/text.js";
import { TokenSymbol } from "../components/token/TokenSymbol.js";
import { StyledButton, StyledDiv } from "../design-system/elements.js";
import { useClipboard } from "../hooks/useCopyClipboard.js";
const pulseAnimation = keyframes `
0% {
  opacity: 1;
  transform: scale(0.5);
}
100% {
  opacity: 0;
  transform: scale(1.5);
}
`;
const WaitingBadge = /* @__PURE__ */ StyledDiv(() => {
    const theme = useCustomTheme();
    return {
        display: "flex",
        alignItems: "center",
        gap: spacing.sm,
        backgroundColor: theme.colors.tertiaryBg,
        border: `1px solid ${theme.colors.borderColor}`,
        padding: `${spacing.md} ${spacing.sm}`,
        borderRadius: radius.lg,
        color: theme.colors.secondaryText,
        fontSize: fontSize.sm,
        fontWeight: 500,
        position: "relative",
        "&::before": {
            content: '""',
            width: "8px",
            height: "8px",
            borderRadius: "50%",
            backgroundColor: theme.colors.accentText,
            animation: `${pulseAnimation} 1s infinite`,
        },
    };
});
/**
 *
 * @internal
 */
export function DepositScreen(props) {
    const activeWallet = useActiveWallet();
    const activeAccount = useActiveAccount();
    const address = activeAccount?.address;
    const { hasCopied, onCopy } = useClipboard(address || "");
    const { connectLocale, client } = props;
    const locale = connectLocale.receiveFundsScreen;
    const isTestnet = props.tx.chain.testnet === true;
    const { data: transactionCostAndData, error: transactionCostAndDataError, isFetching: transactionCostAndDataFetching, refetch: transactionCostAndDataRefetch, } = useTransactionCostAndData({
        transaction: props.tx,
        account: activeAccount,
        supportedDestinations: [],
        refetchIntervalMs: 10_000,
    });
    const theme = useCustomTheme();
    const sponsoredTransactionsEnabled = hasSponsoredTransactionsEnabled(activeWallet);
    if (transactionCostAndDataError) {
        return (_jsx(Container, { style: {
                minHeight: "350px",
            }, fullHeight: true, flex: "row", center: "both", children: _jsx(ErrorState, { title: transactionCostAndDataError?.message || "Something went wrong", onTryAgain: transactionCostAndDataRefetch }) }));
    }
    if (!transactionCostAndData) {
        return _jsx(LoadingScreen, {});
    }
    const totalCost = isNativeToken(transactionCostAndData.token) && !sponsoredTransactionsEnabled
        ? transactionCostAndData.transactionValueWei +
            transactionCostAndData.gasCostWei
        : transactionCostAndData.transactionValueWei;
    const insufficientFunds = transactionCostAndData.walletBalance.value < totalCost;
    const requiredFunds = transactionCostAndData.walletBalance.value
        ? totalCost - transactionCostAndData.walletBalance.value
        : totalCost;
    const openFaucetLink = () => {
        window.open(`https://thirdweb.com/${props.tx.chain.id}?utm_source=ub_deposit`);
    };
    return (_jsxs(Container, { p: "lg", children: [_jsx(ModalHeader, { title: "Insufficient funds", onBack: props.onBack }), _jsx(Spacer, { y: "lg" }), _jsxs(Container, { flex: "column", gap: "sm", children: [insufficientFunds && (_jsx("div", { children: _jsxs(Text, { size: "xs", center: true, color: "danger", multiline: true, children: ["You need", " ", formatNumber(Number.parseFloat(toTokens(requiredFunds, transactionCostAndData.decimals)), 5), " ", transactionCostAndData.token.symbol, " to continue"] }) })), _jsxs(Container, { flex: "row", style: {
                            justifyContent: "space-between",
                            padding: spacing.sm,
                            marginBottom: spacing.sm,
                            borderRadius: spacing.md,
                            backgroundColor: theme.colors.tertiaryBg,
                            border: `1px solid ${theme.colors.borderColor}`,
                        }, children: [activeAccount && (_jsx(WalletRow, { address: activeAccount?.address, iconSize: "md", client: client })), transactionCostAndData.walletBalance.value !== undefined &&
                                !transactionCostAndDataFetching ? (_jsxs(Container, { flex: "row", gap: "3xs", center: "y", children: [_jsx(Text, { size: "xs", color: "secondaryText", weight: 500, children: formatTokenBalance(transactionCostAndData.walletBalance, false) }), _jsx(TokenSymbol, { token: transactionCostAndData.token, chain: props.tx.chain, size: "xs", color: "secondaryText" })] })) : (_jsx(Container, { flex: "row", gap: "3xs", center: "y", children: _jsx(Skeleton, { width: "70px", height: fontSize.xs }) }))] })] }), _jsx(WalletAddressContainer, { onClick: onCopy, children: _jsxs(Container, { flex: "column", gap: "md", center: "both", expand: true, children: [_jsx(Container, { flex: "row", center: "x", children: _jsx(QRCode, { qrCodeUri: address, size: 250, QRIcon: activeWallet && (_jsx(WalletImage, { id: activeWallet.id, size: iconSize.xl, client: client })) }) }), _jsxs(Container, { flex: "row", center: "x", gap: "xs", children: [_jsx(Text, { color: "primaryText", size: "md", children: address ? shortenAddress(address) : "" }), _jsx(CopyIcon, { text: address || "", tip: "Copy address", hasCopied: hasCopied })] })] }) }), _jsx(Spacer, { y: "md" }), _jsx(Text, { multiline: true, center: true, balance: true, size: "sm", className: "receive_fund_screen_instruction", children: locale.instruction }), _jsx(Spacer, { y: "md" }), insufficientFunds ? (_jsxs(WaitingBadge, { children: ["Waiting for funds on ", transactionCostAndData.chainMetadata.name, "..."] })) : (_jsx(Button, { variant: "accent", onClick: props.onDone, fullWidth: true, children: "Continue" })), insufficientFunds && isTestnet && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "md" }), _jsx(Button, { variant: "link", onClick: openFaucetLink, fullWidth: true, children: _jsxs(Container, { flex: "row", center: "x", gap: "xs", color: "accentText", children: [_jsx(CoinsIcon, { size: iconSize.sm }), _jsx(Text, { size: "xs", color: "accentText", weight: 500, center: true, children: "Get testnet funds" })] }) })] }))] }));
}
const WalletAddressContainer = /* @__PURE__ */ StyledButton((_) => {
    const theme = useCustomTheme();
    return {
        all: "unset",
        width: "100%",
        boxSizing: "border-box",
        cursor: "pointer",
        padding: spacing.md,
        display: "flex",
        justifyContent: "space-between",
        border: `1px solid ${theme.colors.borderColor}`,
        borderRadius: radius.lg,
        transition: "border-color 200ms ease",
        "&:hover": {
            borderColor: theme.colors.accentText,
        },
    };
});
